### PR TITLE
Add opencensus support

### PR DESCRIPTION
**Pull request by lanzafame**
@hsanjuan I will also need to go through the fun of `gxing` the `opencensus` pkg and it's dependencies. But I will do that once you have given this PR the green light.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -208,6 +212,25 @@ func (server *Server) handle(s *streamWrap) error {
 func (s *service) svcCall(sWrap *streamWrap, mtype *methodType, svcID ServiceID, ctxv, argv, replyv reflect.Value) error {
 	function := mtype.method.Func
 
+	startT := time.Now()
+	var cctx context.Context
+	var ok bool
+	cctx, ok = ctxv.Interface().(context.Context)
```
Comment (hsanjuan) [inline server.go]: Why not doing this code block inside `handle()` where you have the original `ctx`? That way you wouldn't need to re-cast into the original type.

Perhaps worth separating into it's own helper function.

**Comment by hsanjuan** [comment]
@lanzafame build problem in tests

File: server.go [hunk 2]
```diff
@@ -201,13 +214,16 @@ func (server *Server) handle(s *streamWrap) error {
 	}()
 
 	// Call service and respond
-	return service.svcCall(s, mtype, svcID, ctxv, argv, replyv)
+	return service.svcCall(ctx, s, mtype, svcID, ctxv, argv, replyv)
 }
 
 // svcCall calls the actual method associated
-func (s *service) svcCall(sWrap *streamWrap, mtype *methodType, svcID ServiceID, ctxv, argv, replyv reflect.Value) error {
+func (s *service) svcCall(ctx context.Context, sWrap *streamWrap, mtype *methodType, svcID ServiceID, ctxv, argv, replyv reflect.Value) error {
```
Comment (hsanjuan) [inline server.go]: if you carry `ctx`, then maybe we don't need to pass `ctxv` as well and we can do the `ctxv := reflect.ValueOf(ctx)` inside of this? (unless I'm missing something, seems we're essentially passing the same thing twice)

**Comment by hsanjuan** [comment]
I'm thinking a lot about this and it doesn't feel right to modify go-libp2p-gorpc to support tracing like this.

Particularly, tracing options like latencyDistribution or exposed tracing endpoints should not be part of this library. This is a simple, minimalistic rpc library. I don't like that it starts depending on things other than libp2p or that it assumes tracing is necessary and automatically exposes it.

I have checked how `grpc` supports tracing and they don't. However, they have the notion of Interceptors that can be used for that. We can take a simpler but similar approach and provide a method in `Server` and `Client` to allow users provide a  notification function which gets called on requests. i.e:

 `func (s *Server) SetNotify(func(ctx context.Context, service, method string))`

with a context that is cancelled at the end of the request. This can potentially be useful not only for tracing.

All the rest opentracing configuration and setup should just be part of cluster. We could even just start and end traces in the cluster's RPC server on each method and not touch this library at all, but I understand the convenience of making a change here instead of in every RPC method.

Sorry that I come up with this after so long..

**Comment by lanzafame** [comment]
`grpc-go` doesn't necessarily expose stats and tracing but it does do the ferrying of contexts across the rpc boundary so that tracing can propogate, see https://github.com/grpc/grpc-go/blob/master/stats/stats.go

**Comment by lanzafame** [comment]
From looking at how [OC](https://github.com/census-instrumentation/opencensus-go/blob/53d438757d994ef2e7624cad3de6ac36764496fa/plugin/ocgrpc/client.go) and [promgrpc](https://github.com/piotrkowalczuk/promgrpc/blob/master/prometheus.go) hook into grpc, to do a similar things `gorpc` would need to expose a similar `gorpc/stats` package. 

An example of how grpc does the rpc latency stats for a unary RPC call can be seen here: https://github.com/grpc/grpc-go/blob/91c7ef84b56e1ade18d9665df68cb80231d8ca23/server.go#L838-L855. It is also followed up by the tracing they have built-in.

**Comment by hsanjuan** [comment]
Hmm indeed, I missed we need to modify the original context.

If we focus just on `server.go` (like now). Would it work if we call two user-provided methods, one at start, and one at the end of the call, which return new contexts to replace the existing ones? I wouldn't want to have gorpc collect stats, but to facilitate a user of the library to adjust the calls context and to know when the call starts/ends.

**Comment by hsanjuan** [comment]
Overseeded by #34

**Pull request closed without merging**